### PR TITLE
Feat/sign

### PR DESCRIPTION
**Pull request by VyacheslavDudin**

### CODE REVIEW

File: frontend/src/scenes/SignIn/SignIn.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import { Formik, Form, useField } from 'formik';
+import * as Yup from 'yup';
+import './SignUp.sass';
+import { login } from '../../services/authService';
+import { setLocalStorageItem } from '../../helpers/localStorageHelper';
+
+const InputField = ({ label, ...props }: { label: string, name: string, type: string, placeholder?: string }) => {
```
Comment (MoriArchi) [inline frontend/src/scenes/SignIn/SignIn.tsx:8]: You define the same InputField component in 3 different files. 
Please, generalize it, move this definition to separate file and use everywhere you need InputField

File: frontend/src/scenes/SignIn/SignIn.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+export const SignUp = (props: any) => {
+  return (
+    <div className="signUp">
+      <h1 className="text-center p-5">Sign in</h1>
+      <Formik
+        initialValues={{
+          email: '',
+          password: ''
+        }}
+        validationSchema={Yup.object().shape({
```
Comment (MoriArchi) [inline frontend/src/scenes/SignIn/SignIn.tsx:30]: You can define validationSchema in separate file or in the top of this file. 
In such way defenition happens just once, not on every render of SignUp component

File: frontend/src/scenes/SignUp/Test.jsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      <input className={`text-input ${meta.touched && meta.error ? 'invalid-input' : ''}`} {...field} {...props} />
+      {meta.touched && meta.error ? (
+        <div className="error">{meta.error}</div>
+      ) : null}
+    </div>
+  );
+};
+
+export const SignupForm = props => {
+  const a = 5;
```
Comment (MoriArchi) [inline frontend/src/scenes/SignUp/Test.jsx:23]: Unnecessary variable

File: frontend/src/helpers/webApiHelper.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    try {
+      parsedException = await res.json();
+    } catch (err) {
+      //
+    }
+    throw parsedException;
+  }
+}
+
+export default async function callWebApi(args) {
```
Comment (AlexanderTirik) [inline frontend/src/helpers/webApiHelper.js:51]: We already have the api helper. Check `frontend\src\common\helpers\apiHelper.ts`

File: frontend/package.json [hunk 5]
```diff
@@ -55,5 +55,8 @@
       "last 1 firefox version",
       "last 1 safari version"
     ]
+  },
+  "devDependencies": {
+    "@types/yup": "^0.29.4"
```
Comment (AlexanderTirik) [inline frontend/package.json:60]:
```suggestion
    "@types/yup": "0.29.4"
```

**Comment by vlvereta** [comment]
@VyacheslavDudin please, fix comments and we'll proceed further.

**Comment by VyacheslavDudin** [comment]
All comments are fixed

File: frontend/src/common/helpers/localStorageHelper.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+const setLocalStorageItem = (key: string, value: string): void => localStorage.setItem(key, value);
+
+const getLocalStorageItem = (key:string): string | null => localStorage.getItem(key);
+
+export { setLocalStorageItem, getLocalStorageItem };
```
Comment (AlikRakhmonov) [inline frontend/src/common/helpers/localStorageHelper.ts:5]: 1. Call the file as storageHelper
2. The file should contain methods:
setAccessToken: (token: string) => void
getAccessToken: () => string

File: frontend/src/components/InputField/InputField.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { useField } from 'formik';
+import './InputField.sass';
+
+const InputField = ({ label, ...props }: { label: string, name: string, type: string, placeholder?: string }) => {
```
Comment (AlikRakhmonov) [inline frontend/src/components/InputField/InputField.tsx:5]: component should have an interface for the props
use react-bootstrap

File: frontend/src/scenes/SignIn/SignIn.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+          email: Yup.string()
+            .email('Email is invalid')
+            .required('Email is required'),
+          password: Yup.string()
+            .min(6, 'Password must be at least 6 characters')
+            .required('Password is required')
+        })}
+        onSubmit={async (values:
```
Comment (AlikRakhmonov) [inline frontend/src/scenes/SignIn/SignIn.tsx:26]: move lambda to a separate method

File: frontend/src/scenes/SignIn/SignIn.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+          { email: string, password: string },
+          { setSubmitting }) => {
+          const { token } = await login(JSON.stringify(values));
```
Comment (AlikRakhmonov) [inline frontend/src/scenes/SignIn/SignIn.tsx:29]: Do stringify in service

File: frontend/src/scenes/SignUp/SignUp.sass [hunk 10]
```diff
@@ -0,0 +1,6 @@
+.signUp
+  height: 100vh
+  background-color: #E9ECEF
+
+
+
```
Comment (AlikRakhmonov) [inline frontend/src/scenes/SignUp/SignUp.sass:6]: remove empty lines

File: frontend/src/scenes/SignUp/actions.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { ADD_NEW_USER } from './actionTypes';
+import { ActionCreatorsMapObject } from 'redux';
+
+const addNewUserAction = (userData: object) => ({
```
Comment (AlikRakhmonov) [inline frontend/src/scenes/SignUp/actions.ts:4]: we user sagas and routines, no actions and action types

File: frontend/src/services/authService.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import api from '../common/helpers/apiHelper';
+
+export const login = async (userData: string) => {
```
Comment (AlikRakhmonov) [inline frontend/src/services/authService.ts:3]: userData should be an a object (interface)

File: frontend/src/services/authService.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  const response: Response = await api.post('/api/auth/login', {userData});
+
+  return response.json();
+};
+
+
+export const registration = async (userData: string) => {
```
Comment (AlikRakhmonov) [inline frontend/src/services/authService.ts:10]: userData should be an a object (interface)

File: frontend/src/common/helpers/storageHelper.ts [hunk 14]
```diff
@@ -1 +1,7 @@
-export const getAccessToken = () => localStorage.getItem('token');
+const tokenFieldName = 'token';
```
Comment (vlvereta) [inline frontend/src/common/helpers/storageHelper.ts:1]: Better to create enum **localStorageFields** with fields like Token = 'token' and then use it for methods.

File: frontend/src/common/models/formik/ValSchema.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
+import * as Yup from 'yup';
```
Comment (vlvereta) [inline frontend/src/common/models/formik/ValSchema.ts:1]: Rename file to **ValidationSchemas**

File: frontend/src/common/models/formik/ValSchema.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    .required('Email is required'),
+  password: Yup.string()
+    .min(6, 'Password must be at least 6 characters')
+    .required('Password is required')
+});
+
+const signUpValSchema = Yup.object().shape({
+  fullName: Yup.string()
+    .required('Full Name is required')
+    .max(40, 'Full name should be no longer 40 characters'),
```
Comment (vlvereta) [inline frontend/src/common/models/formik/ValSchema.ts:15]: We have restrictions 3 < n < 100 for name

File: frontend/src/common/models/signIn-signUp/user.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import { IUser } from "../user/user";
```
Comment (vlvereta) [inline frontend/src/common/models/signIn-signUp/user.ts:1]: Rename folder to user or auth

File: frontend/src/common/models/store.ts [hunk 18]
```diff
@@ -1,7 +1,7 @@
-import { IUserState } from '../../reducers/user';
 import { RouterState } from 'connected-react-router';
+import { IUser } from './user/user';
 
 export interface IAppState {
   router: RouterState;
-  user: IUserState;
+  user: IUser;
```
Comment (vlvereta) [inline frontend/src/common/models/store.ts:6]: Should be IUserState with fields: data: IUser, loading: boolean..

File: frontend/src/common/models/user/user.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+interface IUser {
+  id: string;
+  email: string;
+  fullName: string;
+  displayName: string;
+  imageUrl: string | null | undefined;
+  title: string | null | undefined;
+  isLoading: boolean;
```
Comment (vlvereta) [inline frontend/src/common/models/user/user.ts:8]: isLoading should be placed outside of this object

File: frontend/src/components/InputField/InputField.tsx [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+const InputField = ({ label, ...props }: IInputField) => {
+  const style = {
+    width: '300px'
+  };
+  const [field, meta] = useField(props);
+  return (
+    <div className="inputField form-group d-flex flex-column" style={style}>
+      <span>{label}</span>
```
Comment (vlvereta) [inline frontend/src/components/InputField/InputField.tsx:18]: Is there react bootstrap component for Input field?

File: frontend/src/components/InputField/InputField.tsx [hunk 21]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { useField } from 'formik';
+
+interface IInputField {
```
Comment (vlvereta) [inline frontend/src/components/InputField/InputField.tsx:4]: IProps

File: frontend/src/reducers/user.ts [hunk 22]
```diff
@@ -1,25 +1,36 @@
 import { Routine } from 'redux-saga-routines';
-import { fetchUserRoutine } from '../routines/user';
+import { addNewUserRoutine, fetchUserRoutine } from '../routines/user';
```
Comment (vlvereta) [inline frontend/src/reducers/user.ts:2]: Should be fixed according to the proper interface _(without loading field)_.

File: frontend/src/scenes/SignIn/index.tsx [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    const { email, password } = values;
+    const mappedValues = {
+      email,
+      password
+    };
+    const payload = {
+      payload: mappedValues
+    };
+    fetchUser(payload);
+    setSubmitting(false);
```
Comment (vlvereta) [inline frontend/src/scenes/SignIn/index.tsx:27]: Why do you need this?

File: frontend/src/scenes/SignIn/index.tsx [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+const SignIn: FunctionComponent<IProps> = ({ fetchUser }) => {
+  const onSubmit = async (values: IUserInput,
+    { setSubmitting }: { setSubmitting: CallableFunction }) => {
+    const { email, password } = values;
+    const mappedValues = {
+      email,
+      password
+    };
+    const payload = {
+      payload: mappedValues
```
Comment (vlvereta) [inline frontend/src/scenes/SignIn/index.tsx:24]:
```suggestion
      payload: { email, password }
```

File: frontend/src/scenes/SignIn/index.tsx [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+const SignIn: FunctionComponent<IProps> = ({ fetchUser }) => {
+  const onSubmit = async (values: IUserInput,
+    { setSubmitting }: { setSubmitting: CallableFunction }) => {
+    const { email, password } = values;
+    const mappedValues = {
+      email,
+      password
+    };
```
Comment (vlvereta) [inline frontend/src/scenes/SignIn/index.tsx:22]: It's excess.

File: frontend/src/scenes/SignIn/styles.module.sass [hunk 26]
```diff
@@ -0,0 +1,3 @@
+.signIn
+  height: 100vh
+  background-color: #E9ECEF
```
Comment (vlvereta) [inline frontend/src/scenes/SignIn/styles.module.sass:3]: Set and use colors from main colors file.

File: frontend/src/scenes/SignUp/index.tsx [hunk 27]
```diff
@@ -0,0 +1,88 @@
+import React, { FunctionComponent } from 'react';
```
Comment (vlvereta) [inline frontend/src/scenes/SignUp/index.tsx:1]: See comments for Sign In form..

File: frontend/src/scenes/SignUp/styles.module.sass [hunk 28]
```diff
@@ -0,0 +1,3 @@
+.signUp
+  height: 100vh
+  background-color: #E9ECEF
```
Comment (vlvereta) [inline frontend/src/scenes/SignUp/styles.module.sass:3]: Color.

File: frontend/src/services/authService.ts [hunk 29]
```diff
@@ -0,0 +1,38 @@
+import api from '../common/helpers/apiHelper';
+import { IUser } from '../common/models/user/user';
+import { ISignServerResponse } from '../common/models/signIn-signUp/user';
+
+interface ISignUpFields {
```
Comment (vlvereta) [inline frontend/src/services/authService.ts:5]: These interfaces should be centralised.

**Comment by VyacheslavDudin** [comment]
Done!

File: frontend/src/common/helpers/storageHelper.ts [hunk 30]
```diff
@@ -1 +1,7 @@
...
+
+const setAccessToken = (token: string): void => localStorage.setItem(tokenFieldName, token);
+
+const getAccessToken = (): string | null => localStorage.getItem(tokenFieldName);
```
Comment (vlvereta) [inline frontend/src/common/helpers/storageHelper.ts:5]:
```suggestion
const setAccessToken = (token: string): void => localStorage.setItem(LocalStorageFields.Token, token);

const getAccessToken = (): string | null => localStorage.getItem(LocalStorageFields.Token);
```

File: frontend/src/reducers/user.ts [hunk 31]
```diff
@@ -1,32 +1,45 @@
...
-const initialState: IUserState = {
+const initialState: IUserInitState = {
   isLoading: false,
   isAuthorized: false
 };
 
-export default (state = initialState, action: Routine<any>) => {
-  switch (action.type) {
+const reducer = (state = initialState, { type, payload }: Routine<any>) => {
+  switch (type) {
```
Comment (vlvereta) [inline frontend/src/reducers/user.ts:15]: Failed actions?

File: frontend/src/reducers/user.ts [hunk 32]
```diff
@@ -1,32 +1,45 @@
 import { Routine } from 'redux-saga-routines';
-import { fetchUserRoutine } from '../routines/user';
+import { addNewUserRoutine, fetchUserRoutine } from '../routines/user';
 
-export interface IUserState {
+export interface IUserInitState {
```
Comment (vlvereta) [inline frontend/src/reducers/user.ts:4]: It should be in the global models and used for Store typing

File: frontend/src/sagas/user.ts [hunk 33]
```diff
@@ -14,8 +19,23 @@ function* watchUserRequest() {
   yield takeEvery(fetchUserRoutine.TRIGGER, fetchUserRequest);
 }
 
+function* addNewUserRequest({ payload }: any): Routine<any> {
+  try {
+    const { token, user }: ISignServerResponse = yield call(registration, payload);
+    yield put(addNewUserRoutine.success({ payload: user }));
+    setAccessToken(token);
+  } catch (error) {
+    throw Error;
```
Comment (vlvereta) [inline frontend/src/sagas/user.ts:28]: Failed routine should be

File: frontend/src/services/authService.ts [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+  return response.json();
+};
+
+export const registration = async (userInput: ISignUpFields): Promise<IUser> => {
+  const { email, password, fullName } = userInput;
+  const mappedFields = {
+    email,
+    password,
+    fullName
+  };
```
Comment (vlvereta) [inline frontend/src/services/authService.ts:20]: ??

File: frontend/src/styles/colors.sass [hunk 35]
```diff
@@ -2,3 +2,4 @@
   --user-detector-online: yellowgreen
   --user-detector-offline: grey
   --user-detector-border: black
+  --auth-form-background: #E9ECEF;
```
Comment (vlvereta) [inline frontend/src/styles/colors.sass:5]: Remove semicolon

File: frontend/src/reducers/user.ts (hunk 32 above)
Comment (VyacheslavDudin) [reply frontend/src/reducers/user.ts:4]: Maybe remove IUserInitState and use IUserState instead? And initialState will be { isLoading:  false, isAuthorized: false, data: null } of type IUserState

**Pull request merged**